network/address.go

```diff
@@ -275,22 +275,28 @@ func ExactScopeMatch(addr Address, addrScopes ...Scope) bool {
 	return false
 }
 
-// SelectAddressBySpaceNames picks the first address from the given slice that has
-// the given space name associated.
-func SelectAddressBySpaceNames(addresses []Address, spaceNames ...SpaceName) (Address, bool) {
+// SelectAddressesBySpaceNames filters the input slice of Addresses down to
+// those in the input space name.
+func SelectAddressesBySpaceNames(addresses []Address, spaceNames ...SpaceName) ([]Address, bool) {
+	if len(spaceNames) == 0 {
+		logger.Errorf("addresses not filtered - no spaces given.")
+		return addresses, false
+	}
+
+	var selectedAddresses []Address
 	for _, addr := range addresses {
 		if spaceNameList(spaceNames).IndexOf(addr.SpaceName) >= 0 {
-			logger.Debugf("selected %q as first address in space %q", addr.Value, addr.SpaceName)
-			return addr, true
+			logger.Debugf("selected %q as a address in space %q", addr.Value, addr.SpaceName)
+			selectedAddresses = append(selectedAddresses, addr)
 		}
 	}
 
-	if len(spaceNames) == 0 {
-		logger.Errorf("no spaces to select addresses from")
-	} else {
-		logger.Errorf("no addresses found in spaces %s", spaceNames)
+	if len(selectedAddresses) > 0 {
+		return selectedAddresses, true
 	}
-	return Address{}, false
+
+	logger.Errorf("no addresses found in spaces %s", spaceNames)
```

Comment:
do we want something like "falling back to %s" ?

we might want to use the nice formatting from network for space names that brings them in as a nice list, instead of just [foo bar baz] which doesn't mean much to system operators.

I also wonder about Errorf vs Warningf. 
Does this need human intervention right now? I suppose maybe.
(Warning is intended as something that you should file a bug, Error is something that a human should be responding to.)

Reply:
Should it actually be a Debug message?

These methods are used in different contexts. This particular outcome should be reported if required, higher in the stack. The fact that none of the addresses are in any of the input spaces could be operationally valid...

Reply:
If we are returning a list of addresses, I don't think higher-in-the-stack would notice that nothing matched the space list. Which would mean we would need to return [] here, and then the caller can do an len(filteredAddresses) == 0 to notice this.
Given the internals were already dropping Error notices when this happened, I'm guessing the expectation is that all callers won't like/handle an empty list of addresses.

Anyway, we've lived with what's there, we can continue to do so, just thinking about it.